### PR TITLE
Savefig return bytes

### PR DESCRIPTION
**Pull request by prinsherbert**
Returning a bytestring instead of writing to a file allows users to pass the result to different types of encoders, for example to encode a png as base64, or parsers, for example to do post-processing on a svg using BeautifulSoup or just embedding the svg in an HTML-page.

Of course, this is also possible via the `io.BytesIO` file-like object. However, this results in some boilerplate-code.

This PR is as a replacement for the encode_as I suggested in https://github.com/matplotlib/matplotlib/pull/6400 , but now I hope this implementation reduces "API bloat and cognitive overload", as @efiring justly mentioned, to an acceptable level.

The aforementioned PR also suggested the subplot-iterator. I hope to come back to this as well in a different PR, as I also have some ideas to improve this.

### CODE REVIEW

**Comment by prinsherbert** [comment]
I can't seem to solve the problems which are image comparison ones. When running the tests on my machine I notice these warning:

	/home/herbert/Repositories/matplotlib/build/lib.linux-x86_64-2.7/matplotlib/tests/test_streamplot.py:55: MaskedArrayFutureWarning: setting an item on a masked array which has a shared mask will not copy the mask and also change the original mask array in the future.
	Check the NumPy 1.11 release notes for more information.
	  U[:20, :20] = np.nan
	........................................../home/herbert/Repositories/matplotlib/build/lib.linux-x86_64-2.7/matplotlib/font_manager.py:1314: UserWarning: findfont: Font family [u'sans-serif'] not found. Falling back to DejaVu Sans
	  (prop.get_family(), self.defaultFamily[fontext]))
	/home/herbert/Repositories/matplotlib/build/lib.linux-x86_64-2.7/matplotlib/font_manager.py:1324: UserWarning: findfont: Could not match :family=DejaVu Sans:style=normal:variant=normal:weight=400:stretch=normal:size=14.0. Returning /home/herbert/Repositories/matplotlib/build/lib.linux-x86_64-3.5/matplotlib/mpl-data/fonts/ttf/STIXGeneralBolIta.ttf
	  UserWarning)
	/home/herbert/Repositories/matplotlib/build/lib.linux-x86_64-2.7/matplotlib/font_manager.py:1324: UserWarning: findfont: Could not match :family=DejaVu Sans:style=normal:variant=normal:weight=700:stretch=500:size=14.0. Returning /home/herbert/Repositories/matplotlib/build/lib.linux-x86_64-3.5/matplotlib/mpl-data/fonts/ttf/STIXGeneralBolIta.ttf
	  UserWarning)
	/home/herbert/Repositories/matplotlib/build/lib.linux-x86_64-2.7/matplotlib/font_manager.py:1314: UserWarning: findfont: Font family [u'sans serif'] not found. Falling back to DejaVu Sans
	  (prop.get_family(), self.defaultFamily[fontext]))
	/home/herbert/Repositories/matplotlib/build/lib.linux-x86_64-2.7/matplotlib/font_manager.py:1324: UserWarning: findfont: Could not match :family=DejaVu Sans:style=italic:variant=normal:weight=750:stretch=500:size=14.0. Returning /home/herbert/Repositories/matplotlib/build/lib.linux-x86_64-3.5/matplotlib/mpl-data/fonts/ttf/STIXGeneralBolIta.ttf
	  UserWarning)
	/home/herbert/Repositories/matplotlib/build/lib.linux-x86_64-2.7/matplotlib/font_manager.py:1324: UserWarning: findfont: Could not match :family=DejaVu Sans:style=normal:variant=normal:weight=200:stretch=500:size=14.0. Returning /home/herbert/Repositories/matplotlib/build/lib.linux-x86_64-3.5/matplotlib/mpl-data/fonts/ttf/STIXGeneralBolIta.ttf
	  UserWarning)
	/home/herbert/Repositories/matplotlib/build/lib.linux-x86_64-2.7/matplotlib/font_manager.py:1324: UserWarning: findfont: Could not match :family=DejaVu Sans:style=normal:variant=normal:weight=500:stretch=100:size=14.0. Returning /home/herbert/Repositories/matplotlib/build/lib.linux-x86_64-3.5/matplotlib/mpl-data/fonts/ttf/STIXGeneralBolIta.ttf

of matplotlib falling back to fonts. This is strange, as the documentation states that tests only use fonts included in matplotlib :(

**Comment by jenshnielsen** [comment]
That is usually a result of an invalid font cache. Try removing the font cache files from the matplotlib configuration directory

File: lib/matplotlib/__init__.py [hunk 1]
```diff
@@ -1526,6 +1526,7 @@ def _jupyter_nbextension_paths():
     'matplotlib.tests.test_rcparams',
     'matplotlib.tests.test_scale',
     'matplotlib.tests.test_simplification',
+    'matplotlib.tests.test_savefig_return_bytes',
```
Comment (Kojoley) [inline lib/matplotlib/__init__.py]: I think it is more logical correct to place these tests in `test_figure`

**Comment by Kojoley** [comment]
I believe that it will be more clear not to implement `return_bytes` kwarg, just to return bytes if there is no filename was specified.

**Comment by prinsherbert** [comment]
@Kojoley That would change the interface more significantly though

File: lib/matplotlib/figure.py [hunk 2]
```diff
@@ -1662,7 +1663,22 @@ def savefig(self, *args, **kwargs):
...
+        """
+
+        if 'return_bytes' in kwargs and kwargs['return_bytes'] is True:
+            if len(args) > 0:
+                raise ValueError("fname should not be specified if "
+                                 "return_bytes is set True")
+            del(kwargs['return_bytes'])
+            in_memory_file = io.BytesIO()
+            self.savefig(in_memory_file, **kwargs)
+            in_memory_file.seek(0)
```
Comment (Kojoley) [inline lib/matplotlib/figure.py]: There is no need in seeking if you use `getvalue`

**Comment by Kojoley** [comment]
> @Kojoley That would change the interface more significantly though

I do not think so, because you can always pass `None` as a filename.

**Comment by prinsherbert** [comment]
It is a possibility, I explicitly implemented the `return_bytes` to not risk backward-compatibility issues of whatever sort.

**Comment by prinsherbert** [comment]
I am getting these types of errors:

	======================================================================
	FAIL: matplotlib.tests.test_savefig_return_bytes.test_return_bytes
	----------------------------------------------------------------------
	Traceback (most recent call last):
	  File "/home/herbert/.virtualenvs/matplotlib/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
	    self.test(*self.arg)
	  File "/home/herbert/Repositories/matplotlib/build/lib.linux-x86_64-3.5/matplotlib/tests/test_savefig_return_bytes.py", line 47, in test_return_bytes
	    image_as_bytes)
	  File "/home/herbert/Repositories/matplotlib/build/lib.linux-x86_64-3.5/matplotlib/tests/test_savefig_return_bytes.py", line 33, in compare_bytes_with_image
	    '(RMS %(rms).3f)' % err)
	matplotlib.testing.exceptions.ImageComparisonFailure: images not close: /home/herbert/Repositories/matplotlib/result_images/test_savefig_return_bytes/plot_to_png_bytes.png vs. /home/herbert/Repositories/matplotlib/lib/matplotlib/tests/baseline_images/test_savefig_return_bytes/plot_to_png_bytes.png (RMS 5.950)

	----------------------------------------------------------------------

Even right after copying the resulting image to the baseline folder and re-executing with the same command. Is there some non-determinism in matplotlib font rendering? As the letters in the text seem to shift a bit around (mostly horizontally) across runs.

**Comment by tacaswell** [comment]
Changing the return type based on a kwarg value / input type is something we should avoid.

How about putting a function `fig_as_bytes(fig, **save_fig_kwargs)` in `cbook`?  This will not expand the api on `Figure` or `pyplot`, will not add any strange input dependence on return types, and will not require any funny checking on the length of `*args` (I am not sure what other args are meant to be passed through, there may be some corner case where that is essential).  

It also has the feature that it should be easy to vendor if people want to backport it to work with old version of mpl.

**Comment by Kojoley** [comment]
Maybe `fig.to_bytes()` (like `int.to_bytes()`)?

_Offtopic: The `savefig` method name is a bit strage for OOP (`fig.savefig` and `figure.savefig` looks verbose)._

**Comment by prinsherbert** [comment]
> Maybe fig.to_bytes() (like int.to_bytes())?

What I notice is that good and reasonable suggestions done on PR's that have pros and cons. It makes me wonder how a definitive decision (accept, alternative, reject) is made.

**Comment by dopplershift** [comment]
Generally, we look to achieve consensus; @tacaswell has final say.

💯  on the functionality. I'm 👍  on a method for `Figure` rather than a function. `to_bytes()` doesn't quite seem right since this isn't a direct correspondence of the representation of the figure (unlike `int`); maybe `write_bytes()`?

**Comment by QuLogic** [comment]
`write_bytes` seems to imply that it's some I/O, but it's not really. But I'm still not convinced this is any better than accepting `io.BytesIO` (which is already implemented.)

**Comment by Kojoley** [comment]
> `write_bytes` seems to imply that it's some I/O, but it's not really. But I'm still not convinced this is any better than accepting io.BytesIO (which is already implemented.)

I am of the same opinion.

**Comment by dopplershift** [comment]
Well, it *is* going through all of the I/O paths, just not disk-based I/O--it's serializing the figure to an in-memory buffer. It's not a cheap operation.

**Comment by Kojoley** [comment]
In the future the actual logic can be implemented directly in backends (for memory and speed improvements), and then there will be no more I/O

**Comment by QuLogic** [comment]
But `write` implies it goes _to_ something; e.g., `write`, `fwrite`, `fdwrite`, etc. all write to some I/O, either a file, socket, `BytesIO` object, etc. Returning the thing as a result is the exact opposite of that.

I guess there is an argument to be made that just like passing a file _name_ is a shortcut for passing a pre-opened file, this PR is a shortcut for passing `io.BytesIO`. So maybe passing `fname=None` to `savefig` could be that shortcut? I'm just worried that might be confusing for new users.

**Comment by Kojoley** [comment]
>  So maybe passing `fname=None` to `savefig`

I have made the same suggestion:

> I believe that it will be more clear not to implement return_bytes kwarg, just to return bytes if there is no filename was specified.

But @tacaswell against it:

> Changing the return type based on a kwarg value / input type is something we should avoid.

**Comment by efiring** [comment]
If this is going to go through at all, then I think that the idea of using None as the filename to switch to bytes output is the way to go, and I think that @tacaswell can be convinced of this.  It is not uncommon to have a function return None or some object, depending on its inputs.  It is quite logical that if the filename is explicitly None, then the function returns bytes.  Note that I am advocating `fig.savefig(None, **kw)` and not `fig.savefig(**kw)`.  This will yield considerably simpler code than in the present PR, and will be easy to document clearly.

**Comment by efiring** [comment]
Also if this is to go forward, I think the tests can be simplified.  It is not clear to me that an image comparison is needed at all, and certainly not separate comparisons for pyplot and Figure.  If an image comparison is to be included, then it can be done using the standard image_comparison decorator, and writing the bytes out to a file.

**Comment by prinsherbert** [comment]
>  If an image comparison is to be included, then it can be done using the standard image_comparison decorator, and writing the bytes out to a file.

`image_comparison` wants to save the figure itself, so I would not know how to compare a bytestring with an image this way.

**Comment by efiring** [comment]
Write the bytestring out to a file, so it would be the same as if `savefig` had been called with that file name.

**Comment by prinsherbert** [comment]
> Write the bytestring out to a file, so it would be the same as if savefig had been called with that file name.

Yes, and then the figure is overwritten bij savefig being called by the decorator with a string fname argument, hence the bytestring would not be tested, right?

**Comment by efiring** [comment]
You're right, I didn't look closely enough at what the decorator does.
Testing is good, but there should be some proportionality.  Adding 59 lines of testing and two new images for a tiny and nearly bullet-proof modification of a method and a function strikes me as out of proportion, so I would back off to a minimal smoke test.  If there is to be an image test, I would use an existing image, not add two new ones.  But I think the image testing really is overkill.  Nothing in this PR affects image generation and rendering--it is just a matter of what is done to the rendered image.

**Comment by prinsherbert** [comment]
> Nothing in this PR affects image generation and rendering--it is just a matter of what is done to the rendered image.

On the other hand, all code ever written and submitted for production is assumed to be and stay correct. Testing existing plots might work well though.

One example where things go wrong might be if someone decides to implement specific `fname is None` behaviour in pyplot.savefig, overlooking what is in figure.savefig behavior and introducing some bug due to interference.

**Comment by efiring** [comment]
I'm still highly skeptical about the need for an image test; it doesn't make sense to me.  As for pyplot vs Figure, just test in pyplot; that's enough.  Test code has to be maintained just like the rest, and is just as subject to bugs.  It is important to keep it trim and efficient.  It's a matter of balance.

**Comment by tacaswell** [comment]
I am convinced that `fig.savefig(None, **kwargs)` is the right API for this (despite my earlier comments).  Most backends raise an exception if you pass in `None`, it is un-ambiguous what is intended and does not extend the API.

Turns out @mdboom already made this design design choice in the png backend (b3e9443b02162e0fe11fa0bc32ec72c094a713ee) (and I merged it!).

**Comment by prinsherbert** [comment]
Any help with fixing the tests in whatever form would be appreciated though ;)

**Comment by tacaswell** [comment]
The problem is probably due to the two merges from master that are in this branch.

Please see my branch with the same name where I cherry-picked out the non-merge commits.  I suggest that (once you verify that I have not lost any work) you force-push my branch to your branch to update this PR.  Once you do that I think the tests might pass again.

**Comment by prinsherbert** [comment]
I didn't rewrite the `return_bytes` to `fname==None`, as I was awaiting the outcome of this discussion.

But dod so now and force pushed it with @tacaswell's branch. Should arrive here any minute.

**Comment by prinsherbert** [comment]
I have no idea why the new force-pushed code is not showing up here, since it is visible here: https://github.com/prinsherbert/matplotlib/commits/savefig-return-as-bytes

Any clues?

**Comment by NelleV** [comment]
@prinsherbert you forced pushed on the wrong branch: savefig-return-as-bytes instead of savefig-return-bytes.

**Comment by tacaswell** [comment]
@prinsherbert are you abandoning this?

**Comment by prinsherbert** [comment]
@tacaswell I had mayor issues creating tests, managing git and over time lost my environment and forgot my workflow. I just don't feel like putting in more effort, so I hacked a personal version of matplotlib in my 'local' pypi.

The community is free to pick it up though.

**Pull request closed without merging**